core/authn.go

```diff
@@ -34,26 +36,54 @@ type tokenResult struct {
 
 func (a *apiAuthn) handler(next http.Handler) http.Handler {
 	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
-		err := a.auth(req)
-		if err != nil {
-			errorFormatter.Write(req.Context(), rw, err)
+		token, err0 := a.tokenAuth(req)
+		ctx := req.Context()
+		if err0 == nil && token != "" {
+			// if this request was successfully authenticated, pass the token along
+			ctx = context.WithValue(ctx, "token", token)
```

Comment:
In the ACL library I wrote for Square, we hoisted all of the credentials out of the context/environment into a specific "credential" object (which otherwise lived in the context/environment):

https://github.com/square/rails-auth/blob/master/lib/rails/auth/x509/middleware.rb#L43